### ISSUE TITLE
Merge new contexts code

### ISSUE THREAD

**Issue by fredrik-johansson** [mpmath/mpmath#210]
```
Here is a big patch merging changes from the mp4 branch to fully support
multiple contexts (both classes and instances). Also there are some
optimizations and minor bugfixes from that branch. It also adds a machine
precision context, which is still in a very rough state, but which supports
the basics.

A nifty example:

>>> mp.dps = 1000
>>> timing(lambda: mp.findroot(mp.cos, 1))
0.025849207091962804
>>> timing(lambda: mp.findroot(mp.cos, fp.findroot(fp.cos, 1)))
0.016556294165877894

All old tests pass, with the exception of some trivial printing related
issues, and pickling (there is a horrible hack in place atm to fix pickling
-- this needs to be done properly).

This needs more cleanup, and I'd like to do a bit renaming and
reorganizing. Some features that should be shared for the context
implementations is currently duplicated or hackish. A couple of
function/module names are awful and/or awfully inconsistent. There are also
barely any fp tests. We should fix the tests and doctests to import and use
mp and fp objects explicitly, and test both.

Besides the pickling issue, there are also probably issues with generating
the docs now that more things are created dynamically... I haven't looked
at it, but the problems are probably easy to fix.

Also, Vinzent, your code (optimization and matrices) perhaps needs a closer
look. The force_type stuff should be possible to remove; we should always
just convert the inputs, and the fp context will support float/complex. I
did a bit of work on this already, but it needs to be done thoroughly.

Vinzent - do you think it's ok to apply this to the trunk as it is?
Although it will introduce some minor breakage at this point, tests
basically pass, and I'd rather merge as soon as possible so I can also work
on some overdue features and bugfixes without having to worry about having
two codebases.
```

Original issue for #210: http://code.google.com/p/mpmath/issues/detail?id=170
Original author: https://code.google.com/u/111502149103757882156/
Original owner: https://code.google.com/u/111502149103757882156/
